Dockerfile

```diff
@@ -18,59 +18,109 @@ RUN apk --no-cache add \
     && rm -rf /var/cache/apk/*
 
 # Setup repo
-COPY . ${GOPATH}/src/code.gitea.io/gitea
 WORKDIR ${GOPATH}/src/code.gitea.io/gitea
 
+COPY Makefile .
+
+# Download Golang Modules
+COPY go.mod .
+COPY go.sum .
+
+RUN make deps-backend
+
+# Download NPM Packages
+COPY package.json .
+COPY package-lock.json .
+
+RUN make deps-frontend
+
+COPY . .
+
 # Checkout version if set
 RUN if [ -n "${GITEA_VERSION}" ]; then git checkout "${GITEA_VERSION}"; fi \
  && make clean-all build
 
 # Begin env-to-ini build
 RUN go build contrib/environment-to-ini/environment-to-ini.go
 
-# Copy local files
-COPY docker/root /tmp/local
-
-# Set permissions
-RUN chmod 755 /tmp/local/usr/bin/entrypoint \
-              /tmp/local/usr/local/bin/gitea \
-              /tmp/local/etc/s6/gitea/* \
-              /tmp/local/etc/s6/openssh/* \
-              /tmp/local/etc/s6/.s6-svscan/* \
-              /go/src/code.gitea.io/gitea/gitea \
-              /go/src/code.gitea.io/gitea/environment-to-ini
-RUN chmod 644 /go/src/code.gitea.io/gitea/contrib/autocompletion/bash_autocomplete
-
-FROM docker.io/library/alpine:3.18
+FROM docker.io/library/alpine:3.18 AS gitea-base
 LABEL maintainer="maintainers@gitea.io"
 
-EXPOSE 22 3000
-
 RUN apk --no-cache add \
     bash \
     ca-certificates \
-    curl \
     gettext \
     git \
+    curl \
+    gnupg \
+  && rm -rf /var/cache/apk/*
+
+RUN addgroup -S -g 1000 git
+
+FROM gitea-base AS gitea-rootless
```

Comment:
How does this build work with the two tags in the same file? Would `target: gitea` in `docker/build-push-action` skip over these steps for `gitea-rootless`? I assume not, right?

Reply:
You can see in the docker-dryrun workflow, that the first build-push-action build gitea-base and gitea. Then in the next build-push-action you can see the gitea-base layers have all been cached and only the gitea-rootless layers are run.

If not specifying the target when building, docker build will export the final image "as the one to run".